cabal-install/Distribution/Client/ProjectPlanOutput.hs

```diff
@@ -157,3 +175,454 @@ encodePlanAsJson distDirLayout elaboratedInstallPlan elaboratedSharedConfig =
+    -- And not if it didn't need rebuilding in the first place.
+    buildAttempted buildStatus _buildOutcome
+      | not (buildStatusRequiresBuild buildStatus)
+      = False
+
+    -- And not if it was skipped due to a dep failing first.
+    buildAttempted _ (Left BuildFailure {buildFailureReason})
+      | DependentFailed _ <- buildFailureReason
+      = False
+
+    -- Otherwise, succeeded or failed, yes the build was tried.
+    buildAttempted _ (Left BuildFailure {}) = True
+    buildAttempted _ (Right _)              = True
+
+    lookupBuildStatusRequiresBuild def ipkgid =
+      case Map.lookup ipkgid pkgBuildStatus of
+        Nothing          -> def -- Not in the plan subset we did the dry-run on
+        Just buildStatus -> buildStatusRequiresBuild buildStatus
+
+    packagesBuildLocal =
+      selectPlanPackageIdSet $ \pkg ->
+        case pkg of
+          InstallPlan.PreExisting _     -> False
+          InstallPlan.Installed   _     -> False
+          InstallPlan.Configured srcpkg -> elabLocalToProject srcpkg
+
+    packagesBuildInplace =
+      selectPlanPackageIdSet $ \pkg ->
+        case pkg of
+          InstallPlan.PreExisting _     -> False
+          InstallPlan.Installed   _     -> False
+          InstallPlan.Configured srcpkg -> elabBuildStyle srcpkg
+                                        == BuildInplaceOnly
+
+    packagesAlreadyInStore =
+      selectPlanPackageIdSet $ \pkg ->
+        case pkg of
+          InstallPlan.PreExisting _ -> True
+          InstallPlan.Installed   _ -> True
+          InstallPlan.Configured  _ -> False
+
+    selectPlanPackageIdSet p = Map.keysSet
+                             . Map.filter p
+                             $ InstallPlan.toMap plan
+
+
+
+updatePostBuildProjectStatus :: Verbosity
+                             -> DistDirLayout
+                             -> ElaboratedInstallPlan
+                             -> BuildStatusMap
+                             -> BuildOutcomes
+                             -> IO PostBuildProjectStatus
+updatePostBuildProjectStatus verbosity distDirLayout
+                             elaboratedInstallPlan
+                             pkgsBuildStatus buildOutcomes = do
+
+    -- Read the previous up-to-date set, update it and write it back
+    previousUpToDate   <- readPackagesUpToDateCacheFile distDirLayout
+    let currentBuildStatus@PostBuildProjectStatus{..}
+                        = postBuildProjectStatus
+                            elaboratedInstallPlan
+                            previousUpToDate
+                            pkgsBuildStatus
+                            buildOutcomes
+    let currentUpToDate = packagesProbablyUpToDate
+    writePackagesUpToDateCacheFile distDirLayout currentUpToDate
+
+    -- Report various possibly interesting things
+    debugNoWrap verbosity $
+        "packages definitely up to date: "
+     ++ displayPackageIdSet (packagesDefinitelyUpToDate
+          `Set.intersection` packagesBuildInplace)
```

Comment:
Oh otherwise we would end up listing loads and loads of uninteresting deps from the store.